server/src/main/java/org/opensearch/search/DefaultSearchContext.java

```diff
@@ -963,27 +997,58 @@ public BucketCollectorProcessor bucketCollectorProcessor() {
         return bucketCollectorProcessor;
     }
 
+    /**
+     * Evaluate the concurrentSearchMode for the cluster. If the `cluster.search.concurrent_segment_search.mode` setting
+     * is not explicitly set, the evaluation falls back to the `cluster.search.concurrent_segment_search.enabled` boolean
+     * setting which will evaluate to true or false. This is then evaluated to "all" or "none" respectively
+     * @return one of "none, auto, all"
+     */
+    private String evaluateConcurrentSearchMode(Executor concurrentSearchExecutor) {
+        if ((clusterService != null) && concurrentSearchExecutor != null) {
+            String concurrentSearchMode = indexService.getIndexSettings()
+                .getSettings()
+                .get(
+                    IndexSettings.INDEX_CONCURRENT_SEGMENT_SEARCH_MODE.getKey(),
+                    clusterService.getClusterSettings().getAsStringIfSet(CLUSTER_CONCURRENT_SEGMENT_SEARCH_MODE)
+                );
+            if (concurrentSearchMode != null) {
+                return concurrentSearchMode;
+            }
+
+            // new setting is not set, fallback to old settings
+            boolean concurrentSearchSetting = indexService.getIndexSettings()
+                .getSettings()
+                .getAsBoolean(
+                    IndexSettings.INDEX_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey(),
+                    clusterService.getClusterSettings().get(CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING)
+                );
+            return concurrentSearchSetting ? CONCURRENT_SEGMENT_SEARCH_MODE_ALL : CONCURRENT_SEGMENT_SEARCH_MODE_NONE;
+        }
+        return CONCURRENT_SEGMENT_SEARCH_MODE_NONE;
+    }
+
+    @Override
+    public String getConcurrentSearchMode() {
+        return concurrentSearchMode;
+    }
+
     /**
      * Evaluate based on cluster and index settings if concurrent segment search should be used for this request context
      * @return true: use concurrent search
      *         false: otherwise
      */
-    private boolean evaluateConcurrentSegmentSearchSettings(Executor concurrentSearchExecutor) {
+    private boolean evaluateConcurrentSegmentSearchSettings() {
```

Comment:
This method isn't really evaluating concurrent search settings anymore, why can't the logic for `evaluateConcurrentSearchMode` and `evaluateConcurrentSegmentSearchSettings` be in the same method?

Reply:
I wanted to keep the check for system indices separate from computing the `concurrentSearchMode` because they essentially do different things.
Earlier the `evaluateConcurrentSegmentSearchSettings` was actually not evaluating the settings if it was system index but the setting was also boolean so seemed ok, here the setting isn't boolean so it doesn't make sense to combine it. I could evaluate the mode to "none" if it is a system index; but i dont really have a strong preference on that.